### ISSUE TITLE
Performance issue for large data / kotlinx-io

### ISSUE THREAD

**Issue by westnordost** [pdvrieze/xmlutil#217]
**TLDR**: Library user experiences performance issues and requests support for kotlinx-io streaming primitives.

---

As one part of making an app multiplatform, I replaced Java HttpUrlConnection + XML pull parsing with ktor-client + xmlutil + kotlinx-serialization. Unfortunately, the performance at least on parsing large data (tested with ~10MB) got 3-6 times worse. (see [StreetComplete#5686](https://github.com/streetcomplete/StreetComplete/pull/5686) for a short analysis and comparison).

I suspect that using xmlutil's streaming parser instead of deserializing the whole data structure with kotlinx-serialization might improve this somewhat because theoretically, xmlutil should be able start reading the bytes as they come through the wire and in a separate thread than the thread that receives the bytes. (But ultimately not knowing the internals, I can only guess. If you have any other suspicions for the cause of the performance issue, let me know!)

The documentation is a bit thin on streaming, but I understand I need to call `xmlStreaming.newGenericReader(Reader)` to get an `XmlReader` which is a xml pull parser interface. However, I need to supply a (Java) `Reader`, so currently it seems there is no interface for stream parsing on multiplatform.

I understand that byte streaming support and consequently text streaming built on top of that is a bit higgledy-piggledy right now in the Kotlin ecosystem because a replacement for `InputSteam` etc. has never been available in the Kotlin standard library. So, every library that does something with IO implemented their own thing, if anything at all - sometimes based on `okio`, sometimes an own implementation.

However, now it seems like things are about to get better: Both [`ktor`](https://github.com/ktorio/ktor/pull/4032) and apparently [`kotlinx-serialization`](https://github.com/Kotlin/kotlinx.serialization/pull/2707) are being migrated to use [`kotlinx-io`](https://github.com/Kotlin/kotlinx-io) and thus the common interfaces like `Sink`, `Source` and `Buffer`, which are replacements for `InputStream` et al.

So, in case you'd agree that most likely my performance issue with large data stirs from the lack of XML streaming, I guess my request would be to move with kotlinx-serialization and ktor to support a common interface for streaming bytes and text.

**Comment by westnordost** [comment]
I did two [experiments](https://github.com/streetcomplete/StreetComplete/pull/5686#issuecomment-2168085139):

### 1. decoding from stream instead

In my code, I replaced 
`xml.decodeFromString<ApiOsm>(string)`
... with ...
`xml.decodeFromReader<ApiOsm>(xmlStreaming.newReader(inputStream.bufferedReader())`

However, it takes about the **same time**, actually is marginally slower than from string. Also, my assumption or hope that the parser can start parsing the incoming bytes while the http client is receiving them in parallel is apparently wrong. No idea why, but that would be on ktor developers, anyway.

### 2. writing a parser manually, using xmlutil's XmlReader

A manually written parser brings the desired performance gain, cutting the parsing time to about a third. It is still a bit slower than before, but not by much.

Streaming from a stream of bytes is (again) about 25% slower than streaming from a string. When taking into account the time it takes to stream the bytes into a string, it amounts to taking exactly the same time. (And in experiment one I already observed that ktor apparently doesn't make available the input stream until it has been completely received...)

---

In conclusion, streaming doesn't solve anything (for my use case) and for some reason either `kotlinx-serialization` or xmlutil's implementation seems to be quite slow.

**Comment by pdvrieze** [comment]
The generic parser on Jvm supports reading directly from InputStreams (with encoding detection) using the `KtXmlReader` factory function. If you want to use a "native" (stax) reader, you can just create this and get an xml reader calling `StAXReader(myReader)`.

I've been having an initial stab at supporting kotlinx.io which is finally in an usable state. However it is held up a bit because kotlinx.io doesn't support encodings (and I haven't had time to have a decent stab at dealing with that added complexity).

As to performance, I would very much welcome a good performance test suite. As I haven't looked at it explicitly I'm not surprised that it is not optimal. Given all that it does it would never match a hand-rolled parser, but I expect there are quite some optimization options (hence the need to create the test suite to test it - I may repurpose the xmlschema test suite for that - I already include it in the work in progress xmlschema branch - it is big).

**Comment by westnordost** [comment]
I am not able to completely follow. Is this about `newReader(Reader)` vs `newGenericReader(Reader)`? I haven't understood the difference between the two.

> Given all that it does it would never match a hand-rolled parser

Right. The experiment above was made with your xmlutil XmlParser, though, so it is not completely hand-rolled. 

> kotlinx.io doesn't support encodings 

You mean, any to string decoding, or just anything else than UTF-8?

**Comment by pdvrieze** [comment]
The difference between newReader and newGenericReader is that newGeneric reader will always return a KtXmlReader instance, and thus behave the same on each platform (modulo unsupported encodings). NewReader on JVM/Android/Browser will have different implementations.

What I meant with kotlinx.io decoding is that it doesn't support the byte to character conversion (charset decoding/encoding).

Btw. I've had a look at performance. I've done some testing on parsing xml schema (I already had that going, just needed to create the performance test). I've cut of some big chunks of time already, but there is probably a lot more that can be optimized. The main issue seems to be in building the descriptors (and if you want speed, don't put in order restrictions)

**Comment by westnordost** [comment]
> What I meant with kotlinx.io decoding is that it doesn't support the byte to character conversion (charset decoding/encoding).

There's `public fun Source.readString(byteCount: Long): String` and `Sink.writeString(string: String, startIndex: Int = 0, endIndex: Int = string.length)` (since 0.3.5) for UTF-8 strings

https://github.com/Kotlin/kotlinx-io/blob/7c4d095d2fe208abe1eb98a5e07c9b5a95cf3dc8/core/common/src/Utf8.kt#L232

The JVM-implementation includes a `readString` overload where a charset can be defined. (But for my and for most other people's use cases, UTF-8 is probably all they need anyway.)

Though, what most people probably want is a way to encapsulate that in the way the `Reader` does it on Java and it would make sense to add such a class for kotlinx-io directly rather than leave this up for implementation by every single parser/serializer lib that does something with text.
I am not acquainted at all with kotlinx-io yet, so I am not sure if they maybe already have or plan something like that, otherwise it would make sense to create an issue ticket over there.

> Btw. I've had a look at performance. I've done some testing on parsing xml schema (I already had that going, just needed to create the performance test). I've cut of some big chunks of time already

🥳 , that's great! On kotlinx-serialization in particular or even in the core?

**Comment by pdvrieze** [comment]
I have already some UTF8 support in the library as it is needed for native serialization (it actually has a FileInputStream implementation). The Source.readString and Sink.writeString are not that helpful for me as the infrastructure reads by character. Again, I have UTF* code already in the library, but need to move it and create a way for it to work with charsets.

As to performance, this is mainly in the way descriptors (that drive the serialization) are created (there was way to much "convenient" code in there). I haven't actually done anything on the xml parsing itself, only serialization. I've already cut the time used to parse 10000 xmlschema documents in half (in 10 iterations, following a warmup run - warmup is significant here).

**Comment by pdvrieze** [comment]
I have done quite some optimization work on KtXmlReader (shaving at least 30% off the parsing time on the xml schema test suite - from approx 270ms to 200ms). The deserialization has also been optimized, and while it was in the 12000 ms range, it is now at approx 270ms for a list of xml events, or 478ms for both). And it should be worth noting that at least half of the parsing time is taken up by deflating from the resource jar that is used for the testing. I've also added a flag `unchecked` to the deserialization options that removes some correctness checks.

**Comment by westnordost** [comment]
Awesome!

**Comment by westnordost** [comment]
I do expect that and additional most significant speed boost on parsing large XML data in **context of multiplatform** (no Java) will arrive when xmlutil allows to parse from kotlinx-io's `Source` (and HTTP libraries like Ktor will provide a `Source` for the response body) because then parsing can take place in parallel to reading the data from disk.

**Comment by pdvrieze** [comment]
If you want to take a look, I've pushed the "optimization" branch with this work.

**Comment by ComBatVision** [comment]
Hi. When do you plan to release this optimization?

**Comment by pdvrieze** [comment]
I'll probably move it to dev (and snapshot release) in the next few days. I'll try to see if I can add kotlinx.io support as well. Then some soaking/testing is probably good as there are quite some detailed changes that could have broken things.

**Comment by ComBatVision** [comment]
Hi, I had no chance to test night build yet.
Anybody tested this fix?
When will it be in beta release?

**Comment by pdvrieze** [comment]
@ComBatVision I'm building the beta now (there were still some bugs hitting the test suite, there are possibly further bugs/regressions present).

**Comment by ComBatVision** [comment]
@pdvrieze Thank you. I will test it when beta will be available on maven central.

**Comment by pdvrieze** [comment]
It is there now.

**Comment by ComBatVision** [comment]
Should I specify some special parameters or it is by default?

**Comment by ComBatVision** [comment]
This update requires Kotlin 2.0.0 We will test it little bit later. Our project was not migrated to 2.0 yet.

**Comment by westnordost** [comment]
I did a small test. I can confirm a speed boost of about 30% with the new beta.

**Comment by pdvrieze** [comment]
> This update requires Kotlin 2.0.0 We will test it little bit later. Our project was not migrated to 2.0 yet.

I may be able to build a version that uses the older kotlin and serialization. At least when not beta.

**Comment by pdvrieze** [comment]
> I did a small test. I can confirm a speed boost of about 30% with the new beta.

I terms of speed it is important to retain the format across the runs as that caches the building of the file structure. Warmup costs are still quite large compared to further runs. I optimised mainly on repeated runs, rather than the first one.

**Comment by ComBatVision** [comment]
> > This update requires Kotlin 2.0.0 We will test it little bit later. Our project was not migrated to 2.0 yet.
> 
> I may be able to build a version that uses the older kotlin and serialization. At least when not beta.

It will be good to support Kotlin 1.9.24 still, becasue migration to 2.0 on production is still to risky and to complex.
We have Compose in the project and migration is not trivial.

**Comment by westnordost** [comment]
Already 0.90 is built with Kotlin 2.0. Can you link to a source why migration to 2.0 would be risky and complex?

**Comment by pdvrieze** [comment]
I've been doing some work on my own system and found some bugs. In any case I will be doing a release candidate (just to let people know that there may be unforeseen regressions)

**Comment by pdvrieze** [comment]
Btw. As to a 1.9 version, I created a fork, but it failed some tests and I will need to fix them

**Comment by hfhbd** [comment]
Do you still have any plans to natively integrate kotlinx.io? Maybe as a separate artifact?

**Comment by pdvrieze** [comment]
@hfhbd In principle yes. It should work as extension as the existing interface already works with readers/writers/appendables. The thing is that the mapping needs to be done, including handling charset conversions (the hanger).

**Comment by pdvrieze** [comment]
I've just created this in the dev channel (it is currently building). You should be able to access it as snapshot release. This adds extension functions: `encodeToSink` and `decodeFromSource`. They should be efficient, except for JavaScript, where decoding just converts the source to string. On non-JVM targets only UTF-8/UTF-16LE/UTF-16BE/UTF32 are supported.

**Comment by westnordost** [comment]
Now that ktor-client, the http library I use, finally added functions to convert their `ByteReadChannel` data structure to `Source`, I'd like to try it out to test the performance difference (and to already implement it as it will surely be included in an upcoming release :-) )

But anyway, where would I find those snapshot releases?

**Comment by pdvrieze** [comment]
@westnordost I'll try to release a release candidate soon (there are some challenges still with multithreaded correctness). But the snapshots should be in the readme: https://github.com/pdvrieze/xmlutil?tab=readme-ov-file#add-repository

**Comment by westnordost** [comment]
Yeah, I saw that, but I don't understand it 😢. OK, the snapshots are in a different maven repo, but how do I know which version number I should put in the `implementation("io.github.pdvrieze.xmlutil:core:<????????>")` statement in the `build.gradle.kts`?

Multithreaded correctness? Intriguing. What is done on multiple threads in xmlutil? Or you mean, when parsing several XML files in different threads at the same time?

**Comment by pdvrieze** [comment]
The version to use is "0.90.4-SNAPSHOT" (see https://s01.oss.sonatype.org/content/repositories/snapshots/io/github/pdvrieze/xmlutil/core-jvmcommon/ for the list of versions)

**Comment by westnordost** [comment]
I would have expected these functions, 
```
fun IXmlStreaming.newReader(source: Source): XmlReader
fun IXmlStreaming.newWriter(sink: Sink): XmlWriter
```
analogous to the methods with `InputStream`/`OutputStream` on JVM only, but I don't see any. What's the interface? (I definitely got 0.90.4-SNAPSHOT )

**Comment by westnordost** [comment]
I did a quick test, for parsing large XML data, this will be definite speed improvement, unsurprisingly :-)

The dependency name is a bit odd, this is the first time I've seen uppercase dependency names: `io.github.pdvrieze.xmlutil:coreKXIO`

Jetbrains for example uses simply `io` as a suffix for the package that provide interop with kotlinx-io for their json serialization library:

`org.jetbrains.kotlinx:kotlinx-serialization-json-io`

**Comment by pdvrieze** [comment]
The name is a good catch. It should have been called core-io (and will be in the non-snapshot version). Btw. this version should also have a fixed caching that doesn't blow the memory anymore (proper LRU cache).

**Comment by westnordost** [comment]
Caching, yeah, I guess this is the thing that you need to make threadsafe. I do wonder, what is it that is being cached?

**Comment by pdvrieze** [comment]
> Caching, yeah, I guess this is the thing that you need to make threadsafe, that you mentioned. I do wonder, what is it that is being cached?

Basically the format creates shadow descriptors that determine how things are serializer and deserializer - all based on the configuration/policy. This structure is independent of the actual data and is what is cached as some of the logic is not very fast (probably the field reordering is worst - if used).

**Comment by westnordost** [comment]
So with v0.91.0 released, this can be closed.

**Comment by westnordost** [comment]
By the way, might be worth mentioning the new module `core-io` in the readme.
